### ISSUE TITLE
RFC: New React Context API

### ISSUE THREAD

**Issue by langpavel** [formatjs/react-intl#1106]
### This is RFC

There is upcoming change in React 16.3: [New Version of Context](https://github.com/reactjs/rfcs/blob/master/text/0002-new-version-of-context.md#summary)

Is there any plan for new (*probably major / may be minor*) release of `react-intl` supporting this new API?
 - I believe that some issues can disappear with new API
 - old APIs will probably start throwing deprecation warnings with React 16.4
 - old APIs will probably be removed in React 17

**Comment by iamstarkov** [comment]
> I believe that some issues can disappear with new API

For sure, for example `shouldComponentUpdate` wont anymore break language switch

**Comment by iamstarkov** [comment]
i can drive this

**Comment by iamstarkov** [comment]
@okuryu @ericf what do you think about it updating react-intl to a new context api?

**Comment by redonkulus** [comment]
@iamstarkov can it be done in a backwards compatible way?

**Comment by langpavel** [comment]
@redonkulus new Context API is `react>=16.3`, old context will be supported in `react<17`

**Comment by iamstarkov** [comment]
@redonkulus yes, i crafted https://github.com/cssinjs/theming and ensured that it works properly for context updates

**Comment by jantorestolsvik** [comment]
React 16.3 is out. Is this being worked on? I can help if you need.

**Comment by louh** [comment]
Is the idea behind adopting the official Context API that we can dynamically re-render an app when the locale changes, without needing a `key` prop on the `IntlProvider`?

**Comment by jantorestolsvik** [comment]
Yes, and we can rerender only the FormattedMessage etc components and not the whole tree

**Comment by heathzipongo** [comment]
It would be very helpful if `react-intl` would export it's own React Context.Consumer component implementation so that I can access the `intl` object from inside my component without the need for the `injectIntl` HOC.  Please tell me that is also part of the work being done?!

i.e. in my component I can do something like:
```
import {IntlContextConsumer} from 'react-intl'; // Proposed name for the consumer
class MyComponent extends Component {
  render() {
    return (
      <IntlContextConsumer>
        {{intl} => {(
           <input 
             placeholder={intl.formatMessage({id: 'messageId', defaultMessage: 'Enter something'})
             ... // other props
          />
        }}
      </IntlContextConsumer>
    );
  }
}
```

**Comment by langpavel** [comment]
@heathzipongo Yes, this is example case of new context API. I hope this will be supported in the future. And it should be possible to use intl outside of context at your own risc…
But nothing of this is implemented now unfortunately…

**Comment by MarcusHurney** [comment]
Any update on the status of this issue? Also, has anyone upgraded the library to React >= React 16.2 on their own?

**Comment by langpavel** [comment]
I completely rewrote entire library and babel plugin as well and trying write webpack 4 integration with splitting support...
library is much smaller now, it basically working, but unfinished, unpolished, untested, etc.. not ready for publishing yet..

Here is my babel plugin on [npm](https://www.npmjs.com/package/@langpavel/babel-plugin-react-intl), [GitHub](https://github.com/langpavel/babel-plugin-react-intl)

I will publish react-intl as well, but it can take me a month or two… :hourglass_flowing_sand:

**Comment by DragonRaider5** [comment]
This could also fix #58, as the new context API fixes this kind of bugs.

I'll try to do a quick and simple replacement of the old context API and send a PR (/post it here) - let's see wether or not it works out in a feasible amount of time.

**Comment by DragonRaider5** [comment]
Hi, @langpavel that's why I'd like not to do an entire rewrite but instead port the existing package. Depending on how their tests are rewritten, they could be reused - making this a much faster option.

However I'm definitely looking forward to your lib!

**Comment by langpavel** [comment]
Good luck @DragonRaider5 :-)
I originally want create PR, incremental work on top of react-intl… It's possible but you will see, much internal APIs are cumbersome, rich for closures, hacky…
However I'm definitely looking forward to your work too! :-)

**Comment by DragonRaider5** [comment]
I just finished the basic work - everything seems to work fine in my own Project (and now properly updates every formatted element when changing the locale).

Yet there still are some issues:
- the tests (mostly the unit tests) fail, as f.e. there is no `getChildContext()` anymore (another example: a new HOC is used around every Component, breaking all `setState` spies in the unit tests)
- there is no polyfill for the new context API included (and it uses `static getDerivedStateFromProps()`) -> `react >= 16.3.1` is required

Every help is very welcome!

To install:
1. Clone the [fork](https://github.com/DragonRaider5/react-intl/tree/feature/new-context-api)
2. Change to the branch `feature/new-context-api`
3. Navigate there with the command line of your choice
4. Run `npm i` and `npm run build` to build the package
5. Run `npm link` to link the package globally
6. Navigate to the project you want to use [the fork](https://github.com/DragonRaider5/react-intl/tree/feature/new-context-api) in
7. Run `npm link react-intl` (caution: **you need to do this after every single `npm install` in this package**)
8. Use `react-intl` as usual

**Comment by DragonRaider5** [comment]
Are you also still interested in this @MarcusHurney @heathzipongo @jantorestolsvik @louh @iamstarkov ?

Help always is appreciated, check out the post above :)

**Comment by MarcusHurney** [comment]
@DragonRaider5 
I am still interested. We ended up using the create-react-context polyfill, but if yahoo doesn't make this update internally, I think it makes sense to patch it as you've done. Thank you for putting the effort into this! I am not sure we'll use the fork now, but in the future the work you've invested here may be enormously helpful. I'll be in touch and in case we do decide to go with a custom solution, I'd be happy to help you further with this.

**Comment by DragonRaider5** [comment]
@MarcusHurney
Thanks, nice to hear from you.

I'm using my fork in production and there have been no problems yet.

Still, as you noticed, it's far from being finished. One of the big challenges yet to come will be updating the test cases, because of the changed internals (state spies) and the testing using shallow rendering with the new context API.

I might finish the testing, when my project has gone further.
However I'd be happy about any help and even though I might not have the time to actively develop it, I'd like to help everybody thinking about improvements :)

**Comment by vinz51** [comment]
Hello, any news on this feature ? I don't see a pull request for that ?! 😃

**Comment by MarcusHurney** [comment]
Hi team Yahoo! Is this upgrade something you guys have planned? It'd be really helpful to just get a yes or no at minimum! Thanks for your hard work.

**Comment by DragonRaider5** [comment]
Hey @vinz51,
sorry for the late response... I thought I had already done so :C

It's still WIP - I haven't done much on it since I got it working. Tests still not fixed... So that's why there is no PR yet.
However @papasmile reached out to me and we are looking forward to further work on in this.

My latest PR contains a Slack workspace ([react-intl.slack.com](https://react-intl.slack.com)) for further conversation - I'd be happy to see you guys join us there.

**Comment by StorytellerCZ** [comment]
@DragonRaider5 Do you have an invite form for that Slack channel?

**Comment by DragonRaider5** [comment]
Hey,
I finished all the unit tests for the refactoring! Now I'm struggling to get the builds going again...

@okuryu @inukshuk @ericf might any of you be able to help me on this one? I'd like to finally merge the changes, but I don't know lot about rollup.

CC @MarcusHurney @langpavel @jantorestolsvik

**Comment by papasmile** [comment]
cc @redonkulus ^^^

**Comment by redonkulus** [comment]
As I mentioned in another issue, we will not have any resources to update this internally. So I would suggest that the community push this forward. This would mean updating tests, documentation, testing in your own applications.

I haven't looked into the changes in depth, but will this require a major version bump for `react-intl`? Or can the existing context code exist with the new API? I would prefer to not break existing, but I'm not sure. Please let me know. Thanks!

**Comment by DragonRaider5** [comment]
@redonkulus No it won't require a major version bump and will be 100% compatible due to two polyfills and me following the old naming convention for the HOC (it's still exported as `injectIntl`).

I already updated the code, fixed all unit tests & am using it in my own project. I just can't get the CDN rollup builds running again - that's what I hoped to get help from you for, as I'm not familiar with building for CDNs...

**Comment by iamstarkov** [comment]
folks, we better hurry 

<img width="930" alt="screenshot 2018-11-02 at 9 42 52 pm" src="https://user-images.githubusercontent.com/559321/47939619-423ef000-dee8-11e8-9c2a-ab187c07741d.png">

https://reactjs.org/blog/2018/10/23/react-v-16-6.html#deprecations-in-strictmode

**Comment by iamstarkov** [comment]
@DragonRaider5 please share your insights on how your fork has performed?

**Comment by DragonRaider5** [comment]
Well... In my project it works just as it did with the old version - despite me not having problems with locale changes and the update tree anymore.

You can try it out yourself and give me feedback by cloning my [branch](https://github.com/DragonRaider5/react-intl/tree/feature/new-context-api), running `npm i && npm run build && npm link` in the cloned folder and running `npm link react-intl` in the project you're using `react-intl` in.

**Comment by iamstarkov** [comment]
@DragonRaider5 can you publish your version of `react-intl` under your own scope?

**Comment by iamstarkov** [comment]
@DragonRaider5 like DragonRaider5/react-intl

**Comment by DragonRaider5** [comment]
@iamstarkov great idea, thanks! I'll look to do so in the evening and keep you updated here.

**Comment by DragonRaider5** [comment]
Done `npm i @dragonraider5/react-intl` :D

**Comment by DragonRaider5** [comment]
Yes you're right - I got the builds to work again (I had to update and swap out a few rollup plugins after updating).

Todos are:
[ ] Fix tests for `React@0.14.1`
[ ] Rebase to get latest changes from `react-intl/master`

**Comment by scottdickerson** [comment]
I will work on the testcases.

**Comment by StorytellerCZ** [comment]
Hi all, I have made my own version couple months back to get some of the fixes from the PRs.

https://github.com/Freedombase/react-intl/tree/release/v3.0.0

When #1186 came around it quickly overtook my initial efforts on the same problem so I added it as well  and dropped support for React before v16 (since I don't have to worry about backward compatibility in my projects).
The test suite was updated to the latest version as well and I also migrated things like `expect` to the new `jest` version.
I wanted to mention it, just in case that something I have done might be of interest, especially regarding the test cases.

**Comment by DragonRaider5** [comment]
Hey @StorytellerCZ, thanks for your affords.

However I've already done all of this too while updating Babel and Rollup - maintaining support for older React versions.

There are only three broken tests for `react@0.14` remaining. Afterwards it is ready to be merged.


Thanks in advance to @scottdickerson for investigating in those!

**Comment by langpavel** [comment]
It looks like yahoo will be on turn really soon.
I can see @redonkulus is maintainer now. Can we know your opinion?
Thanks!

**Comment by langpavel** [comment]
> will this require a major version bump for `react-intl`?

I'm for **major** version bump — even if new version can work with modern stack, it may not work with older (APIs are same for users of library, but cross library APIs are not)

**Comment by StorytellerCZ** [comment]
@langpavel I agree, major version bump would make sense only if you discontinued support for older versions (which my branch does) or broke some API in older minor version, which is not the case here. So a minor version bump is most appropriate.

**Comment by iamstarkov** [comment]
@DragonRaider5 @scottdickerson do you plan to merge it upstream/here?

**Comment by bensampaio** [comment]
Is there any progress on this? I have a big dependency on this project but if it is being dropped then I think that should be made clear on the README file so that everyone finds an alternative :/

**Comment by redonkulus** [comment]
@bensampaio the most up to date progress is in the PR https://github.com/yahoo/react-intl/pull/1186. Someone needs to fix the tests that are failing for React 0.14.x, maybe you can donate your time to get this to work so we can unblock the PR?

**Comment by bensampaio** [comment]
@redonkulus thanks for your reply. Unfortunately I don't think I would have the time to do that. Still I tried to have a look but I can only find linting problems so I couldn't get an idea of the dimension of the problem. What exactly is missing?

**Comment by robrichard** [comment]
@redonkulus is publishing this as a major version and dropping support for older versions of react out of the question? I would prefer that over loading an additional library (`create-react-context`) as it adds unneeded code to the bundles of applications that are on the latest version of React.

Either way, I think this change should be published as a new major version as there could be consuming code that is accessing `intl` by defining the legacy `contextTypes` static property.

**Comment by bensampaio** [comment]
I was also thinking about that. Why keep support for react 0.14? That only keeps this project from evolving in the right direction. Projects like `react-redux` and `react-router` dropped support for react 0.14 so why not this one? That will also make the maintenance of this project easier if there isn't that much time to pay attention to it anyway.

**Comment by bensampaio** [comment]
@DragonRaider5 I definitely agree and I am not here to make pressure or demands. Just trying to understand the issue and what can be done to simplify it.

Are there no plans for yahoo to move to a more recent version of react? Can't this project drop already support for react 0.14 and yahoo catches up later? Since it will be a major release they don't have to use it yet.

Regarding `create-react-context`, is it always going to be included in the bundle or only when using react 0.14?

Moving away from react 0.14 would make it easier to maintain this project. Besides that, it would allow for returning strings from the ´Formatted*´ components and, if I am not wrong, that would reduce the use of `injectIntl`. This is just how I look at it, maybe I am simplifying it too much. However, I am sure I would be more confident to contribute if there was less to maintain and to be aware of.

In any case, good work everyone! This is a really cool project anyhow. I am just excited to see it move forward (and to reduce the size of my vendors 😁).

**Comment by DragonRaider5** [comment]
As we can't do "optional dependencies" it'll be downloaded to your machine anyways.

However it depends on your bundler whether or not it will be included in your consumer bundle - see:
`index.js (create-react-context):`
```javascript
import React from 'react';
import createReactContext from './implementation';

export default React.createContext || createReactContext;
```

I'd also like to drop support for `react@0.14`, maybe @redonkulus can reconsider it? However we'll still be required to keep the support for `react@15` so the polyfill will stay anyways 🤣

I'd also like to point you to a separate issue regarding cleaner code #1224 👍

**Comment by redonkulus** [comment]
@DragonRaider5 well said on your previous comment about maintenance and expectations; I agree completely.

I take upgrading a major version very seriously and try only do so when absolutely necessary. If there is a way to support both versions without cluttering the codebase, I would prefer that. For this particular issue, it looks like we are able to achieve that successfully. 

I'm sure for most applications and projects, migrating to a new major version is as simple as changing the version for your dependency and running the install command. For us, we have applications on 0.14, 15 and 16. We also have hundreds of modules internally that rely on react-intl and upgrading their major version requires a lot of effort and resources that we do not have to support. This particular enhancement looks like it can be done without needing to upgrade the major version. I do not want to just do it to expedite the release. 

Alternatively, since this code is open source, there is nothing preventing users from forking the code and releasing it with these new Context API changes and dropping support for older versions. I prefer this didn't happen, but if changing to the new Context API is a requirement of your project then that might be the best way forward until the PR is merged.

**Comment by bensampaio** [comment]
@DragonRaider5 and @redonkulus thanks for the replies I get the point.

I'll see how it works with `create-react-context` when it is released, but it seems to work as I would expect so that's cool. Looking forward for the release :)

**Comment by scottdickerson** [comment]
@DragonRaider5 any updates on the rebase from master? We're really looking forward to getting this into our build to cleanup the warnings!

**Comment by DragonRaider5** [comment]
@scottdickerson I finished it, but @redonkulus commented on it (see #1186) and I've not gotten a response from him yet :)

**Comment by redonkulus** [comment]
Thanks for the call out. There are a lot of changes going on in that PR and I would implore others on this thread to review the code and add comments. Additionally, full down the fork/branch and try it out in your own applications to ensure nothing is broken. I do not have intimate knowledge of this codebase (I inherited this from the previous dev), so I would prefer the code review be driven by the community.

**Comment by DragonRaider5** [comment]
You can live test my changes using `npm i @dragonraider5/react-intl`.
However this release isn't up to date (latest rebase is missing), I'm going to update it later.

**Comment by redonkulus** [comment]
Since https://github.com/yahoo/react-intl/pull/1186 was merged. Closing this.
